### PR TITLE
HV: move vm configuration check to pre-build time

### PR DESCRIPTION
**Pull request by jsun26intel**
This patch set will move the VM configuration check to pre-build stage,
a test program will do the check for pre-defined VM configuration
data before making hypervisor binary. If test failed, the make
process will be aborted. So once the hypervisor binary is built
successfully or start to run, it means the VM configuration has
been sanitized.

Tracked-On: #5077

Signed-off-by: Victor Sun <victor.sun@intel.com>
Reviewed-by: Jason Chen CJ <jason.cj.chen@intel.com>
Acked-by: Eddie Dong <eddie.dong@intel.com>

### CODE REVIEW

File: hypervisor/arch/x86/guest/vm.c [hunk 1]
```diff
@@ -774,12 +774,22 @@ void prepare_vm(uint16_t vm_id, struct acrn_vm_config *vm_config)
 void launch_vms(uint16_t pcpu_id)
 {
 	uint16_t vm_id;
+	uint64_t pre_vm_cpu_affinity = 0UL;
 	struct acrn_vm_config *vm_config;
 
 	for (vm_id = 0U; vm_id < CONFIG_MAX_VM_NUM; vm_id++) {
 		vm_config = get_vm_config(vm_id);
 		if ((vm_config->load_order == SOS_VM) || (vm_config->load_order == PRE_LAUNCHED_VM)) {
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vm.c:782]: In this case, will "split this if ( .. == SOS_VM || .. == PRE_LAUNCHED_VM)" to "if (...SOS_VM) .... else {}" simplify the code?

Comment (jsun26intel) [reply hypervisor/arch/x86/guest/vm.c:782]: sure, will update.

**Comment by wenlingz** [comment]
OK to Verify

File: hypervisor/arch/x86/guest/vm.c [hunk 2]
```diff
@@ -774,15 +774,25 @@ void prepare_vm(uint16_t vm_id, struct acrn_vm_config *vm_config)
...
 			}
-
+		} else if (vm_config->load_order == SOS_VM) {
+			/* Deduct pcpus of PRE_LAUNCHED_VMs */
+			vm_config->cpu_affinity = ALL_CPUS_MASK ^ pre_vm_cpu_affinity;
+			if (vm_config->cpu_affinity == 0UL) {
+			/* in case physical cores are not all up as expected */
+				pr_err("no vCPUs for SOS VM (vm id: %d)", vm_id);
+				break;
+			}
```
[the remark this answers is not in the document]
Comment (dongyaozu) [reply hypervisor/arch/x86/guest/vm.c:794]: for (vm_id = 0U; vm_id < CONFIG_MAX_VM_NUM; vm_id++) {
                vm_config = get_vm_config(vm_id);
                if ((vm_config->load_order == SOS_VM) || (vm_config->load_order == PRE_LAUNCHED_VM)) {
                        if (pcpu_id == get_configured_bsp_pcpu_id(vm_config)) {
                                prepare_vm(vm_id, vm_config);
                        }
                }
                if ((vm_config->load_order == SOS_VM && pcpu_id == BSP) {
                      sos_vm_ptr = &vm_array[vm_id];
               }
        }


??

[the remark this answers is not in the document]
Comment (jsun26intel) [reply hypervisor/arch/x86/guest/vm.c:794]: ok, will update.
And acrn-config related code need to be changed too in this PR.

**Comment by jsun26intel** [comment]
@wenlingz In this PR, make of pre_build is changed a lot. please check whether build scripts of Yocto need to be modified.

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br>No New Function Declaration/Definition Mismatch

**Pull request merged**